doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+# PREVIEW: New `libelektra-core` API 2023
+
+## Core API
+
+### Note on names
+
+From the [use cases](/doc/usecases/core/README.md) we know that there is a collection structure which contains elements with a name and value (and metadata).
+
+The current API uses these names for those concepts:
+
+- Collection: `KeySet`
+- Element: `Key`
+- Name of element: —
+- Value of element: —
+
+The API proposed below uses (to keep as much as possible from above):
+
+- Collection: `ElektraKeyset`
+- Element: `ElektraKey`
+- Name of element: `ElektraKeyname`
+- Value of element: `ElektraKeyvalue`
+
+Another (maybe less confusing) option would be:
+
+- Collection: `ElektraSet` or `ElektraMap`
+- Element: `ElektraEntry`
+- Name of element: `ElektraKey` or `ElektraName`
+- Value of element: `ElektraValue`
+
+### Note on public structs
+
+The public structs `ElektraKeyname` and `ElektraKeyvalue` defined below are not problematic or limiting in terms of forward compatibility.
```

Comment:
I disagree. E.g. order or if namespace is separate should be implementation detail (only shared with mmap).

Anyway, we should have a decision about why we have opaque pointers, this is a common question and not so well known.

Reply:
1. Whether namespace is separate or not is IMO actually decision we need to make. IMO (and I've stated that before), namespace and name should be entirely separate things. Now that I have thought more about the API, I actually don't see **any benefit** in storing the namespace as the first byte of the name field. IMO it just complicates things.

2. Yes, field order could be reason to make the struct opaque. But in this case there will always be the same 3 fields, a pointer, a size and a byte (for namespace). I don't see why we would ever want to change the order. In any case, I don't think field order outweighs all the benefits that having these public structs brings.

> we should have a decision about why we have opaque pointers, this is a common question and not so well known.

Then **you** need to write one. You are the only one who has been with Elektra long enough to know the actual reason.